iceoryx_utils/include/iceoryx_utils/cxx/function_ref.hpp

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_UTILS_CXX_FUNCTION_REF_HPP
+#define IOX_UTILS_CXX_FUNCTION_REF_HPP
+
+#include <cstddef>
+#include <memory>
+#include <type_traits>
+
+namespace iox
+{
+namespace cxx
+{
+/// @brief cxx::function_ref is a non-owning reference to a callable.
+///
+///        It has these features:
+///         * No heap usage
+///         * No exceptions
+///         * Stateful lambda support
+///         * C++11/14 support
+/// @code
+///         // Usage as function parameter
+///         void fuu(cxx::function_ref<void()> callback)
+///         {
+///             callback();
+///         }
+///         // Call the lambda
+///         fuu([]{ doSomething(); });
+///
+///         // Usage with l-values
+///         // Pitfall: Ensure that lifetime of callable suits the point in time of calling callback()
+///         auto callable = [&]{ doSomething(); };
+///         cxx::function_ref<void()> callback(callable);
+///         // Call the callback
+///         callback();
+/// @endcode
+
+template <typename SignatureType>
+class function_ref;
+
+template <class ReturnType, class... ArgTypes>
+class function_ref<ReturnType(ArgTypes...)>
+{
+    using SignatureType = ReturnType(ArgTypes...);
+    using SelfType = function_ref<SignatureType>;
+    template <typename T>
+    using EnableIfNotFunctionRef = typename std::enable_if<!std::is_same<std::decay<T>, function_ref>::value>::type;
+
+  public:
+    /// @brief Creates an empty function_ref
+    function_ref() noexcept
```

Comment:
not really a big fan of this. optional<function_ref> would be better in my point of view, but it seems the decision was to do it this way

on the other hand, you would result in the same, invalid object after a move operation